spec/engine_spec.rb

```diff
@@ -0,0 +1,271 @@
+      expect(subject.value_status.values).to eq({})
+      expect(subject.max_depth).to be(nil)
+      expect(subject.graph.nodes).to eq(Set.new)
+    end
+  end
+
+  describe "#add_worker" do
+    it "must add worker class to #worker_classes and WorkerTasks instance to #worker_tasks" do
+      Async { subject.add_worker(TestEngine::TestWorker) }
+
+      expect(subject.instance_variable_get(:@worker_classes)).to eq({ Ronin::Recon::Values::Domain => [TestEngine::TestWorker] })
+      expect(subject.instance_variable_get(:@worker_tasks)[Ronin::Recon::Values::Domain][0]).to be_kind_of(Ronin::Recon::WorkerTasks)
+    end
+  end
+
+  describe "#on" do
+    let(:block) { proc { |foo| puts "foo" } }
+
+    context "for :value event" do
+      it "must pass block to #value_callbacks" do
+        subject.on(:value, &block)
+
+        expect(subject.instance_variable_get(:@value_callbacks)).to eq([block])
+      end
+    end
+
+    context "for :connection event" do
+      it "must pass block to #connection_callbacks" do
+        subject.on(:connection, &block)
+
+        expect(subject.instance_variable_get(:@connection_callbacks)).to eq([block])
+      end
+    end
+
+    context "for :job_started event" do
+      it "must pass block to #job_started_callbacks" do
+        subject.on(:job_started, &block)
+
+        expect(subject.instance_variable_get(:@job_started_callbacks)).to eq([block])
+      end
+    end
+
+    context "for :job_completed event" do
+      it "must pass block to #job_completed_callbacks" do
+        subject.on(:job_completed, &block)
+
+        expect(subject.instance_variable_get(:@job_completed_callbacks)).to eq([block])
+      end
+    end
+
+    context "for :job_failed event" do
+      it "must pass block to #job_failed_callbacks" do
+        subject.on(:job_failed, &block)
+
+        expect(subject.instance_variable_get(:@job_failed_callbacks)).to eq([block])
+      end
+    end
+
+    context "for unsupported event type" do
+      it "must raise ArgumentError" do
+        expect {
+          subject.on(:foo, &block)
+        }.to raise_error(ArgumentError)
+      end
+    end
+  end
+
+  describe "#enqueue_mesg" do
+    before do
+      Async { subject.add_worker(TestEngine::TestWorker) }
+    end
+
+    context "for Message::Value" do
+      let(:mesg) { Ronin::Recon::Message::Value.new(value) }
+
+      it "must add value to #value_status" do
+        Async { subject.enqueue_mesg(mesg) }
+
+        expect(subject.value_status.values).to eq({ value => { TestEngine::TestWorker => :enqueued } })
+      end
+    end
+
+    context "for Message::SHUTDOWN" do
+      let(:mesg) { Ronin::Recon::Message::Value.new(value) }
+
+      it "must equeue message for each WorkerTask" do
+        Async { subject.enqueue_mesg(mesg) }
+
+        expect(subject.instance_variable_get(:@worker_tasks).values[0][0].instance_variable_get(:@input_queue).items).to eq([mesg])
```

Comment:
Maybe a less invasive test would be to add `expect(subject.instance_variable_get(:@worker_tasks)[value.class].first).to receive(:enqueue_message).with(mesg)` before calling `subject.enqueue_mesg(mesg)`.